### ISSUE TITLE
Subclassing seaborn.objects.Plot

### ISSUE THREAD

**Issue by maurosilber** [mwaskom/seaborn#3487]
Hi!

I was considering subclassing `Plot` to add a `__call__` method, as I want to create a declarative plot specification, but provide the data a posteriori via a call to the `Plot` instance:

```python
class Plot(so.Plot):
    def __call__(self, result: Mapping[str, NDArray], /):
        plot = self.plot._clone()
        plot._data.source_data = result
        return plot
```

(By the way, do you see any problem in doing this?)

My issue is that `Plot._clone` and the return of its methods are hardcoded with `Plot`:
https://github.com/mwaskom/seaborn/blob/1f788201822941a3aca7fedf0bcd49d16ebc313b/seaborn/_core/plot.py#L395-L397

Could it be changed to:

```python
 def _clone(self) -> Self: 
     """Generate a new object with the same information as the current spec.""" 
     new = self.__class__()
```

I would not mind preparing a PR for this.

Thank you for this new seaborn interface and all your work!

**Comment by mwaskom** [comment]
> I was considering subclassing Plot to add a __call__ method, as I want to create a declarative plot specification, but provide the data a posteriori via a call to the Plot instance

Whether or not sub-classing Plot in general is a good idea (I'm not sure), your proposed code access a private attribute which you probably shouldn't be doing?

I also don't think I understand the actual use case. Maybe there is a simpler way to achieve what you want...

**Comment by maurosilber** [comment]
I need to be able to declare and export (outside Python) a Plot without actually producing (drawing) the plot, as the data itself might not yet be available.

While I haven't tried with all methods, this seems to work:
```python
import seaborn.objects as so

plot = so.Plot().add(so.Line(), x="time", y="position")
```
as it does not access the data yet.

Then, I could inspect the private attributes encoding the plot (`plot._layers` and others) and export them in the format I need.

If I try to plot, it would raise an error, as `data` was never provided:
```python
plot.show()  # error
```

But when I receive the data, I can assign it to this private attribute and plot:
```python
plot._data.source_data = {"time": ..., "position": ...}
plot.show()  # ok
```

So, yes, I would be using private attributes, which is a bad idea if they change in the future. But it is easier than reimplementing all you have done to build a declarative Plot object.

**Comment by mwaskom** [comment]
Sorry I am not quite following what you mean by "export outside Python".

I am asking because your use-case sounds somewhat similar to other idea I have considered, e.g. a method for defining the common data source + mappings outside of the plot constructor (https://github.com/mwaskom/seaborn/issues/2926). But that would still be "in Python" as it would just be another method on Plot.

I have also played around with thinking about how you might define a Plot in yaml (which I guess could be paired with exporting a Plot to yaml). In that case the data source would probably be pointing at a csv filename? That part of the experiment was always somewhat unclear.

But if the only usecase for a Plot subclass requires other "illegal" operations like modifying private data attributes I'm probably not too positively disposed towards it...

**Comment by maurosilber** [comment]
Yes, by export outside Python I mean to a text format like YAML.

In my use case, it would be to an XML-based format called [SED-ML](https://sed-ml.org/) (Simulation Experiment Description Markup Language), which is used in the systems biology community. The data itself can be provided as a path to a CSV, or as a description of the simulation (integrate this differential equation in such and such way). The community created this declarative interexchange format to be able to reproduce other researchers experiments in different simulation tools.

I neither know nor see any more use cases to "export" or "serialize" a `seaborn.objects.Plot` to other text based formats. If someone would want to reproduce my plot, I would just give them my Python script. Did you have another use case in mind?

Regarding https://github.com/mwaskom/seaborn/issues/2926, from inside Python I could just create a function:
```python
def plot(data):
  return so.Plot(data=data).add(...)
```
The only benefit I can think of a `Plot.assign(data)` would be if creating all the layers and such took a non-negligible amount of time.

(yes, I'm not being very supportive of my use case 😅)

**Comment by maurosilber** [comment]
Closing this. I rather see #2926 implemented (particularly if `so.Plot()` does not take any arguments).
